language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/encoding/ValuePrimitiveEncoding.scala

```diff
@@ -42,6 +42,8 @@ trait ValuePrimitiveEncoding[TC[_]] {
   implicit def valueOptional[A: TC]: TC[P.Optional[A]]
 
   implicit def valueMap[A: TC]: TC[P.Map[A]]
+
+  implicit def valueGenMap[K: TC, V: TC]: TC[P.GenMap[K, V]]
```

Comment:
Finally, the choice of `imm.Map` will variously lead to overlapping instances here, which are rejected with a compiler error. And a good thing too, as they would have incompatible implementations, each wrong for the other. So `P.Map` or `P.GenMap` will need a newtype.

Reply:
We still have overlapping instances with `InsertOrdMap` (Since its implement `Map`).  
Do you think I should use `SortedMap` for `P.Map` ?

Reply:
No, use an unboxed newtype for `immutable.Map` as the value of `type P.TextMap` instead. If you [define `subst`](https://failex.blogspot.com/2017/04/the-high-cost-of-anyval-subclasses.html#markdown-header-when-is-a-label-a-string-when-is-it-not), you can set `val TextMap: ImmutableMapFactory[TextMap] = subst(Map)`.

Reply:
NB: the type parameter to `subst` is kinded like this: `F[C[A, +B] <: Map[A, B] with MapLike[A, C[A, B]]]`; accordingly those are the upper bounds on the abstract `TextMap`.

Reply:
To make this perfectly smooth will require a few other changes:

1. add `extends PrimitiveInstances` to `abstract class Primitive`
2. define empty `abstract class PrimitiveInstances`
3. in `object PrimitiveInstances`, define

```scala
implicit def TextMapCanBuildFrom[K, V]
    : CanBuildFrom[Primitive.TextMap.Coll, (K, V), Primitive.TextMap[K, V]] = ...
```

where implementation derives from `Primitive.TextMap.canBuildFrom` probably by means of some private member of `OnlyPrimitive`.

4. use `-Xsource:2.13` where `map` and friends return the wrong type. We document that we expect Scala codegen users to use this option anyway.

Reply:
`binding` changes in ff923d997a; did not check the remainder of modules.